p2p/src/main/java/bisq/network/p2p/storage/persistence/HistoricalDataStoreService.java

```diff
@@ -0,0 +1,187 @@
+                        return true;
+                    }
+
+                    // Otherwise we only add data if the requesters version is older then
+                    // the version of the particular store.
+                    String storeVersion = entry.getKey();
+                    boolean newVersion = Version.isNewVersion(storeVersion, requestersVersion);
+                    String details = newVersion ?
+                            "As our historical store is a newer version we add the data to our result map." :
+                            "As the requester version is not older as our historical store we do not " +
+                                    "add the data to the result map.";
+                    log.info("The requester had version {}. Our historical data store has version {}.\n{}",
+                            requestersVersion, storeVersion, details);
+                    return newVersion;
+                })
+                .map(e -> e.getValue().getMap())
+                .forEach(result::putAll);
+
+        log.info("We found {} entries since requesters version {}",
+                result.size(), requestersVersion);
+        return result;
+    }
+
+    public Map<P2PDataStorage.ByteArray, PersistableNetworkPayload> getMapOfLiveData() {
+        return store.getMap();
+    }
+
+    public Map<P2PDataStorage.ByteArray, PersistableNetworkPayload> getMapOfAllData() {
+        Map<P2PDataStorage.ByteArray, PersistableNetworkPayload> result = new HashMap<>(getMapOfLiveData());
+        result.putAll(allHistoricalPayloads);
+        return result;
+    }
+
+
+    ///////////////////////////////////////////////////////////////////////////////////////////
+    // MapStoreService
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
+    // TODO optimize so that callers to AppendOnlyDataStoreService are not invoking that often getMap
+    // ProposalService is one of the main callers and could avoid it by using the ProposalStoreService directly
+    // instead of AppendOnlyDataStoreService
+
+    // By default we return the live data only. This method should not be used by domain clients but rather the
+    // custom methods getMapOfAllData, getMapOfLiveData or getMapSinceVersion
+    @Override
+    public Map<P2PDataStorage.ByteArray, PersistableNetworkPayload> getMap() {
+        return store.getMap();
+    }
+
+    @Override
+    protected void put(P2PDataStorage.ByteArray hash, PersistableNetworkPayload payload) {
+        if (anyMapContainsKey(hash)) {
+            return;
+        }
+
+        getMapOfLiveData().put(hash, payload);
+        persist();
+    }
+
+    @Override
+    protected PersistableNetworkPayload putIfAbsent(P2PDataStorage.ByteArray hash, PersistableNetworkPayload payload) {
+        if (anyMapContainsKey(hash)) {
+            return null;
+        }
+
+        PersistableNetworkPayload previous = getMapOfLiveData().put(hash, payload);
```

Comment:
Yes it will be always null, but we override the method signature from MapStoreService and follow the putIfAbsent signature from Map. In our special case it wil be always null. I can change it to make it more clear and add a comment.